### PR TITLE
inductor: Fix non-deterministic gradients in slice_scatter backward fusion

### LINKED ISSUE #188890: Non-deterministic gradients when using torch.compile and PyTorch >= 2.12
### 🐛 Describe the bug

The attached code seems to produce non-deterministic and wildly different gradients when run with `torch.compile` on PyTorch >= 2.12. Code run without `torch.compile`, as well as code run on older PyTorch versions produces identical gradients all the time.

The reproduction script has been reduced with the use of LLM.

```python
#!/usr/bin/env python3
import os

os.environ.setdefault("CUBLAS_WORKSPACE_CONFIG", ":4096:8")

import torch
import torch.nn.functional as F


def forward_fn(inputs, score_weight):
    length, dim = inputs.shape
    scores = torch.einsum("td,gd->gt", inputs, score_weight)
    score_delta = scores[:, None, :] - scores[:, :, None]

    pair_scores = torch.einsum("tn,Tn->tT", inputs, inputs)
    pair_weights = pair_scores[None] * torch.exp(score_delta)
    context = torch.einsum("gtT,Td->gtd", pair_weights, inputs)
    context = F.layer_norm(context, (dim,))

    return context.repeat_interleave(4, dim=-1)[:1, :, :1].view(length, 1)


def main():
    torch.use_deterministic_algorithms(True)
    torch.backends.cuda.matmul.allow_tf32 = True

    print(
        f"torch={torch.__version__} cuda_build={torch.version.cuda} "
        f"gpu={torch.cuda.get_device_name()} runs=2 failure_threshold=1.0"
    )

    torch.manual_seed(1337)
    score_weight = torch.empty(4, 224, device="cuda").normal_(std=0.02).requires_grad_()
    compiled_fn = torch.compile(forward_fn)
    inputs = torch.empty(296, 224, device="cuda").normal_(std=0.02)

    score_weight.grad = None
    with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
        outputs = compiled_fn(inputs, score_weight)
        loss = outputs.sum()
    loss.backward()
    grads_a = score_weight.grad.detach().float().reshape(-1).cpu()
    grads_a_finite = torch.isfinite(grads_a).all().item()
    grad_left_norm = float(torch.linalg.vector_norm(grads_a))

    score_weight.grad = None
    with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
        outputs = compiled_fn(inputs, score_weight)
        loss = outputs.sum()
    loss.backward()
    grads_b = score_weight.grad.detach().float().reshape(-1).cpu()
    grads_b_finite = torch.isfinite(grads_b).all().item()
    grad_right_norm = float(torch.linalg.vector_norm(grads_b))

    grad_rel = float(torch.linalg.vector_norm(grads_b - grads_a)) / max(
        grad_left_norm, 1e-30
    )
    all_grads_finite = grads_a_finite and grads_b_finite
    failed = not all_grads_finite or grad_rel >= 1.0

    print(
        f"failed={failed} all_grads_finite={all_grads_finite} "
        f"max_seen_grad_rel={grad_rel:.8e} "
        f"grad0_norm={grad_left_norm:.8e} "
        f"grad1_norm={grad_right_norm:.8e}"
    )
    raise SystemExit(int(failed))


if __name__ == "__main__":
    main()
```

### Error logs

```
torch=2.12.1+cu132 cuda_build=13.2 gpu=NVIDIA H200 runs=2 failure_threshold=1.0
failed=True all_grads_finite=True max_seen_grad_rel=4.07311368e+00 grad0_norm=2.91221905e+00 grad1_norm=1.24026461e+01
```

### Versions

```
Collecting environment information...
PyTorch version: 2.12.1+cu132
Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (conda-forge gcc 14.3.0-16) 14.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.13.11 | packaged by conda-forge | (main, Dec  6 2025, 11:24:03) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-1053-aws-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to:
GPU models and configuration: GPU 0: NVIDIA H200
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  96
On-line CPU(s) list:                     0-95
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Platinum 8488C
CPU family:                              6
Model:                                   143
Thread(s) per core:                      1
Core(s) per socket:                      48
Socket(s):                               2
Stepping:                                8
BogoMIPS:                                4800.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd ida arat avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid cldemote movdiri movdir64b md_clear serialize amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               4.5 MiB (96 instances)
L1i cache:                               3 MiB (96 instances)
L2 cache:                                192 MiB (96 instances)
L3 cache:                                210 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-47
NUMA node1 CPU(s):                       48-95
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
```

cc @mruberry @kurtamohler @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by Tejas-Raj01**
Fixes #188890

### Description
When `torch.use_deterministic_algorithms(True)` is enabled, the AOTAutograd decomposition of `repeat_interleave` combined with slicing generates `slice_scatter` followed by `sum`. Inductor's scheduler aggressively fuses these, leading to complex strided layouts where the overlap analysis fails. 

This results in the Triton backend generating unsafe `tl.store` operations instead of `atomic_add`, causing a memory race condition and wild gradient mismatches.

**The Fix:**
This PR forces realization on `slice_scatter` nodes when deterministic algorithms are enabled. This prevents unsafe layout fusion with downstream reductions, eliminating the memory race condition on complex strided views and guaranteeing strict deterministic accumulations.

### Testing
- [x] Reproduced the issue locally and verified the exact fix using the user-provided script.
- [x] Passed `lintrunner` locally.
- [x] Passed `pytest test/inductor/test_torchinductor.py -k "deterministic"` locally (18/18 passed).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by Tejas-Raj01** [comment]
### Research & Reasoning for this Fix

**1. The Root Cause (Silent Data Corruption):**
The WAW (Write-After-Write) hazard occurs because AOTAutograd generates `slice_scatter` ops with complex strided layouts (involving `FloorDiv` and `Mod`). The scheduler's current overlap analysis, which relies on `_sympy`'s continuous interval arithmetic, fails to prove injectivity for these discrete integer patterns. Consequently, it falsely assumes no overlap and fuses the kernel using `tl.store` instead of `atomic_add`, leading to corrupted gradients.

**2. Why this Fix? (Trade-off):**
* **Immediate Safety:** Forcing `.realize()` acts as a zero-blast-radius bandage. It enforces a kernel boundary, physically preventing the memory race condition.
* **Deterministic Mode:** By gating this behind `deterministic` flags, we ensure that production/eager performance remains untouched. This mode is primarily for debugging, where correctness is non-negotiable.

**3. Future Durable Fix:**
The ideal fix belongs in the scheduler (`sizevars.py`). I am researching a **Fast-Path Diophantine Pattern Matcher** to mathematically guarantee injectivity for strided patterns (checking index multipliers like $A \ge C \times D$) in $O(1)$ time, which would allow us to remove this `.realize()` boundary entirely in the future.

---
**Changes:**
- Added regression test `test_slice_scatter_deterministic_mutation` in `test_torchinductor.py`.
- Added explanatory comments referencing issue #188890.
- Updated gating to check both `torch.are_deterministic_algorithms_enabled()` and `config.deterministic`.

**Comment by embe-pw** [comment]
This does not look like something that should be dependent on the deterministic setting – the gradients seem to be wrong, not just slightly non-deterministic.

**Comment by jansel** [comment]
@claude review these changes again

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -4155,13 +4068,22 @@ def inner_fn(idx):
...
         inner_fn=inner_fn,
         ranges=list(x.get_size()),
     )
 
+    # [Fix for #188890] Prevent fusion overlapping in strided slice_scatter.
+    # A permanent fix belongs in the scheduler's overlap analysis. For now,
+    # we enforce a kernel boundary to prevent WAW data races and silent
+    # data corruption when determinism is requested.
+    if torch.are_deterministic_algorithms_enabled() or config.deterministic:
+        res.realize()
```
Comment (jansel) [inline torch/_inductor/lowering.py:4083]: How does calling .realize() fix a WAW data race?  This should not matter.

**Review by jansel — CHANGES_REQUESTED** [review]
I don't think this is fixing the root cause of the issue.
